### PR TITLE
Add selector for parallelization backend

### PR DESCRIPTION
**Pull request by ronald-jaepel**
Based on changes made in the PR #38 

This fixes parallelization on windows with iPython and callbacks.

### CODE REVIEW

File: examples/characterize_chromatographic_system/column_transport_parameters.py
Comment (schmoelder) [inline examples/characterize_chromatographic_system/column_transport_parameters.py:1]: Does this now also run again?

Comment (ronald-jaepel) [reply examples/characterize_chromatographic_system/column_transport_parameters.py:1]: On my PC: yes. It works with regular python, with ipython and from within jupyter notebooks.

**Comment by ronald-jaepel** [comment]
Testing Flynns examples I've found a few more necessary changes before we merge

**Comment by schmoelder** [comment]
Thanks for investigating this! :)

**Comment by schmoelder** [comment]
@ronald-jaepel would you mind pushing the WIP so I can test it locally?

**Comment by schmoelder** [comment]
Hey @ronald-jaepel this is looking very promising. I rebased and squashed everything into one commit. 

Next steps:
- [ ] Clean up (e.g. todos)
- [ ] Add unit/integration tests. Maybe we even find a way to test in an ipython environment?

**Comment by schmoelder** [comment]
Another thing we could consider (while we're in here) is to unify all of the `evaluate_` and `evaluate_population` methods. For this purpose, we could extend the `ensures2d` decorator to reshape the corresponding outputs.
Similarly, we could unify `run_post_evaluation_processing` and `run_post_generation_processing`. But maybe this could also be a separate issue...

**Comment by ronald-jaepel** [comment]
Regarding the next steps:

The remaining todo in the parallelizationBackend.py is "Add documentation in optimization chapter". I don't know which "chapter" is meant.

**Comment by schmoelder** [comment]
@ronald-jaepel I pulled and installed ipyparallel. However, when running the test, I get the following error(s):

```
ERROR: test_parallel_cadet_initialization (__main__.Test_ParallelizationSimple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jo/code/CADET-Process/tests/test_parallelization_adapter.py", line 91, in test_parallel_cadet_initialization
    results = backend.evaluate(evaluation_function, [0.01] * 4, n_cores=4)
  File "/home/jo/code/CADET-Process/CADETProcess/optimization/parallelizationBackend.py", line 183, in evaluate
    results = results_que.get()
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 357, in get
    raise e
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 530, in _resolve_result
    results = self._collect_exceptions(results)
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 546, in _collect_exceptions
    return error.collect_exceptions(results, self._fname)
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/error.py", line 248, in collect_exceptions
    raise err
ipyparallel.error.CompositeError: one or more exceptions raised in: evaluation_function
[0:apply]NameError: name 'Cadet' is not defined
[1:apply]NameError: name 'Cadet' is not defined
[2:apply]NameError: name 'Cadet' is not defined
[3:apply]NameError: name 'Cadet' is not defined

======================================================================
ERROR: test_parallelization_backend (__main__.Test_ParallelizationSimple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jo/code/CADET-Process/tests/test_parallelization_adapter.py", line 43, in test_parallelization_backend
    results = backend.evaluate(evaluation_function, [0.01] * 4, n_cores=4)
  File "/home/jo/code/CADET-Process/CADETProcess/optimization/parallelizationBackend.py", line 183, in evaluate
    results = results_que.get()
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 357, in get
    raise e
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 530, in _resolve_result
    results = self._collect_exceptions(results)
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/client/asyncresult.py", line 546, in _collect_exceptions
    return error.collect_exceptions(results, self._fname)
  File "/home/jo/software/miniconda3/envs/dev/lib/python3.10/site-packages/ipyparallel/error.py", line 248, in collect_exceptions
    raise err
ipyparallel.error.CompositeError: one or more exceptions raised in: evaluation_function
[0:apply]NameError: name 'time' is not defined
[1:apply]NameError: name 'time' is not defined
[2:apply]NameError: name 'time' is not defined
[3:apply]NameError: name 'time' is not defined
```
Can you please check again if they work on your side?

Also, please add a test for the parallelization backend in the context of an OptimizationProblem/Optimizer.

**Comment by ronald-jaepel** [comment]
> Also, please add a test for the parallelization backend in the context of an OptimizationProblem/Optimizer.
Which Optimizer should I use? Which one is least likely to be deprecated at some point? NSGA3?

**Comment by schmoelder** [comment]
`U_NSGA3 ` seems like a good idea.

**Comment by ronald-jaepel** [comment]
On my end **if cd'd into tests-folder** the test runs fine on Windows Python 3.10 and WSL Ubuntu Python 3.10 ¯\_(ツ)_/¯
From the root directory: ipyparallel fails. 

I can't figure it out and I don't feel like sinking more time into ipyparallel. I'd suggest dropping support for ipyparallel because joblib and pathos both work fine.

Will start on the optimizationproblem test now.

**Comment by ronald-jaepel** [comment]
Added a simple test copied from the pymoo test. It, however, fails with this error:

````
  File "C:\Users\ronal\src\cadet-process\CADETProcess\optimization\optimizationProblem.py", line 651, in _evaluate_population
    raise CADETProcessError(
CADETProcess.CADETProcessError.CADETProcessError: Cannot use dict cache for multiprocessing.
````

even though use_diskcache is set to false. 

Debugging now, just sharing

**Comment by schmoelder** [comment]
> even though use_diskcache is set to false.

That's the issue. You need to use diskcache. dict(ionary)cache cannot be used for multi-processing as it's not thread-safe.

**Comment by schmoelder** [comment]
> I can't figure it out and I don't feel like sinking more time into ipyparallel. I'd suggest dropping support for ipyparallel because joblib and pathos both work fine.

Sounds reasonable.

**Comment by ronald-jaepel** [comment]
Found a new interaction: if you set n_cores on the optimizer and thereafter set the parallelization_backend like this

```
optimizer.n_cores = 10
optimizer.parallelization_backend = backend
```


then the backend's default of 1 core will be used and the value set for the optimizer will be ignored. Does your rewriting change that behavior?

**Comment by schmoelder** [comment]
Since we decided to drop IPYParallel for now, I'm leaving this here for future reference:

```
class IpyparallelAdapter(ParallelizationBackendBase):
    """Runner using the ipyparallel parallelization backend. Can evaluate a function
    over all individuals in a population-list.
    Attributes
    ----------
    n_cores : Integer >= -1
        Number of cores to be used. n_cores = 0 or n_cores = -1 uses all available cores.
    Methods
    -------
    evaluate
        Evaluates the arg: function at all individuals in population
    """
    _ipyparallel_cluster_file = String()

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        import ipyparallel as ipp
        self._ipp = ipp

    def evaluate(self, function, population, n_cores=None):
        if n_cores is None:
            n_cores = self.n_cores

        if n_cores < -1:
            raise ValueError("ipyparallel backend does not support n_cores of < -1.")
        if n_cores in [-1, 0]:
            n_cores = None

        if self._ipyparallel_cluster_file is None:
            # if no Cluster has been started yet, start a new cluster and connect to it
            ipyparallel_cluster = self._ipp.Cluster(n=n_cores).start_and_connect_sync()
            self._ipyparallel_cluster_file = ipyparallel_cluster.cluster.cluster_file
        else:
            # if a Cluster exists, connect to it
            ipyparallel_cluster = self._ipp.Cluster.from_file(self._ipyparallel_cluster_file).connect_client_sync()

        # get all engines in the cluster
        engines_all = ipyparallel_cluster[:]
        results_que = engines_all.map_async(function, population)

        # comment this line out if you don't like the progress bar over the function evaluations
        results_que.wait_interactive()

        results = results_que.get()
        return results
```

Other than that, I reviewed the PR, cleaned up some things and if the tests run, I'll merge.

**Comment by schmoelder** [comment]
Halt Stop! Before merging, add documentation! :nerd_face:

**Pull request merged**